File: scripts/data_processing/blog_scraper.rb
#!/usr/bin/env ruby

require 'nokogiri'
require 'net/http'
require 'uri'
require 'json'
require 'fileutils'
require 'time'

class BlogScraper
  BASE_URL = 'https://jaroslavlachky.sk'
  BLOG_URL = "#{BASE_URL}/blog/"
  LOGIN_URL = "#{BASE_URL}/sign-in/"
  OUTPUT_DIR = './scraped_data'
  
  def initialize(username = nil, password = nil)
    @scraped_articles = []
    @failed_urls = []
    @username = username
    @password = password
    @cookies = {}
    @authenticated = false
    setup_output_directory
  end
  
  def scrape_all
    puts "Starting to scrape #{BLOG_URL}"
    
    # Login if credentials provided
    if @username && @password
      puts "Attempting to login..."
      if login
        puts "✓ Successfully logged in"
        @authenticated = true
      else
        puts "✗ Login failed, continuing without authentication"
      end
    else
      puts "No credentials provided, scraping public content only"
    end
    
    page = 1
    total_scraped = 0
    
    loop do
      puts "\n--- Scraping page #{page} ---"
      article_urls = scrape_article_urls_from_page(page)
      
      if article_urls.empty?
        puts "No more articles found. Stopping."
        break
      end
      
      puts "Found #{article_urls.length} articles on page #{page}"
      
      article_urls.each do |url|
        begin
          article_data = scrape_article(url)
          if article_data
            @scraped_articles << article_data
            total_scraped += 1
            puts "✓ Scraped: #{article_data[:title][0..50]}..."
            
            # Save individual article
            save_article_to_file(article_data)
            
            # Small delay to be respectful
            sleep(0.5)
          else
            puts "✗ Failed to scrape: #{url}"
            @failed_urls << url
          end
        rescue => e
          puts "✗ Error scraping #{url}: #{e.message}"
          @failed_urls << url
        end
      end
      
      page += 1
      
      # Safety break - adjust if needed
      break if page > 50
    end
    
    puts "\n=== Scraping Complete ==="
    puts "Total articles scraped: #{total_scraped}"
    puts "Failed URLs: #{@failed_urls.length}"
    
    save_summary
    save_failed_urls if @failed_urls.any?
  end
  
  private
  
  def login
    begin
      # First, get the login page to extract form details
      login_page_response = fetch_url_raw(LOGIN_URL)
      return false unless login_page_response
      
      doc = Nokogiri::HTML(login_page_response.body)
      
      # Look for WordPress login form
      form = doc.css('form#loginform, form[action*="wp-login"], form[action*="sign-in"]').first
      if form
        action_url = form['action']
        action_url = "#{BASE_URL}#{action_url}" if action_url&.start_with?('/')
        action_url ||= "#{BASE_URL}/wp-login.php"
      else
        # Fallback to standard WordPress login
        action_url = "#{BASE_URL}/wp-login.php"
      end
      
      # Prepare login data
      login_data = {
        'log' => @username,
        'pwd' => @password,
        'wp-submit' => 'Log In',
        'redirect_to' => BLOG_URL,
        'testcookie' => '1'
      }
      
      # Submit login form
      uri = URI(action_url)
      response = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https') do |http|
        request = Net::HTTP::Post.new(uri)
        request['User-Agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'
        request['Content-Type'] = 'application/x-www-form-urlencoded'
        request['Referer'] = LOGIN_URL
        
        # Add existing cookies
        request['Cookie'] = cookies_to_string if @cookies.any?
        
        request.body = URI.encode_www_form(login_data)
        
        http.request(request)
      end
      
      # Extract cookies from response
      extract_cookies_from_response(response)
      
      # Check if login was successful
      if response.code == '302' || response.code == '200'
        # Verify authentication by checking a protected page
        test_response = fetch_url_raw(BLOG_URL)
        return test_response && !test_response.body.include?('PLATENÝCH ČLENOV')
      end
      
      false
    rescue => e
      puts "Login error: #{e.message}"
      false
    end
  end
  
  def extract_cookies_from_response(response)
    response.get_fields('Set-Cookie')&.each do |cookie|
      cookie_parts = cookie.split(';').first.split('=', 2)
      if cookie_parts.length == 2
        @cookies[cookie_parts[0]] = cookie_parts[1]
      end
    end
  end
  
  def cookies_to_string
    @cookies.map { |k, v| "#{k}=#{v}" }.join('; ')
  end
  
  def setup_output_directory
    FileUtils.mkdir_p(OUTPUT_DIR)
    FileUtils.mkdir_p("#{OUTPUT_DIR}/articles")
  end
  
  def scrape_article_urls_from_page(page_num)
    url = page_num == 1 ? BLOG_URL : "#{BLOG_URL}page/#{page_num}/"
    
    begin
      response = fetch_url(url)
      return [] unless response
      
      doc = Nokogiri::HTML(response)
      
      # Try multiple selectors to find article links
      article_links = []
      
      # Common WordPress selectors
      selectors = [
        'article h2 a',
        '.entry-title a',
        '.post-title a',
        'h2.entry-title a',
        '.article h2 a'
      ]
      
      selectors.each do |selector|
        links = doc.css(selector)
        if links.any?
          article_links = links.map { |link| normalize_url(link['href']) }
          break
        end
      end
      
      # Fallback: look for any links that seem like blog posts
      if article_links.empty?
        all_links = doc.css('a[href]')
        article_links = all_links.map { |link| link['href'] }
                                .select { |href| href && href.include?(BASE_URL) && !href.include?('/blog/') && href.count('/') >= 4 }
                                .map { |href| normalize_url(href) }
                                .uniq
      end
      
      article_links.uniq
    rescue => e
      puts "Error fetching page #{page_num}: #{e.message}"
      []
    end
  end
  
  def scrape_article(url)
    response = fetch_url(url)
    return nil unless response
    
    doc = Nokogiri::HTML(response)
    
    # Extract title
    title = extract_title(doc)
    return nil unless title
    
    # Extract content
    content = extract_content(doc)
    return nil if content.empty?
    
    # Extract metadata
    date = extract_date(doc)
    excerpt = extract_excerpt(doc)
    
    {
      url: url,
      title: title,
      content: content,
      date: date,
      excerpt: excerpt,
      scraped_at: Time.now.iso8601,
      word_count: content.split.length
    }
  end
  
  def extract_title(doc)
    selectors = [
      'h1.entry-title',
      '.entry-title',
      'h1',
      'title'
    ]
    
    selectors.each do |selector|
      element = doc.css(selector).first
      if element
        title = element.text.strip
        return title unless title.empty?
      end
    end
    
    nil
  end
  
  def extract_content(doc)
    selectors = [
      '.pf-content .entry_content',
      '.blog_entry_content',
      '.entry_content',
      '.post-content',
      '.article-content',
      '.content',
      'article .entry-content',
      'main article'
    ]
    
    selectors.each do |selector|
      element = doc.css(selector).first
      if element
        # Remove unwanted elements - including specific Slovak blog elements
        element.css('script, style, nav, header, footer, aside, .comments, .social-share, .in_share_element, .fb-like, .twitter-like, .printfriendly, .ve_form_element, form, .mw_social_icons_container, .related_posts').remove
        
        content = element.text.strip
        return content if content.length > 100  # Minimum content length
      end
    end
    
    # Fallback: try to get main content area
    main_content = doc.css('main, .main, #main, #content, .content-area, .blog-content').first
    if main_content
      main_content.css('script, style, nav, header, footer, aside, .comments, .social-share, .in_share_element, .fb-like, .twitter-like, .printfriendly, form').remove
      content = main_content.text.strip
      return content if content.length > 100
    end
    
    ''
  end
  
  def extract_date(doc)
    selectors = [
      'time[datetime]',
      '.published',
      '.entry-date',
      '.post-date',
      '[class*="date"]'
    ]
    
    selectors.each do |selector|
      element = doc.css(selector).first
      if element
        date_str = element['datetime'] || element.text.strip
        begin
          return Time.parse(date_str).iso8601
        rescue
          # Continue to next selector
        end
      end
    end
    
    nil
  end
  
  def extract_excerpt(doc)
    selectors = [
      '.excerpt',
      '.entry-summary',
      '.post-excerpt'
    ]
    
    selectors.each do |selector|
      element = doc.css(selector).first
      return element.text.strip if element
    end
    
    # Fallback: first paragraph of content
    first_p = doc.css('.entry-content p, .post-content p, .content p').first
    first_p ? first_p.text.strip[0..200] : ''
  end
  
  def fetch_url(url)
    response = fetch_url_raw(url)
    return nil unless response
    
    if response.code == '200'
      response.body
    else
      puts "HTTP #{response.code} for #{url}"
      nil
    end
  end
  
  def fetch_url_raw(url)
    uri = URI(url)
    
    Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == 'https') do |http|
      request = Net::HTTP::Get.new(uri)
      request['User-Agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'
      
      # Add cookies if we have them
      if @cookies.any?
        request['Cookie'] = cookies_to_string
      end
      
      response = http.request(request)
      
      # Extract cookies from response
      extract_cookies_from_response(response)
      
      response
    end
  rescue => e
    puts "Network error for #{url}: #{e.message}"
    nil
  end
  
  def normalize_url(url)
    return nil unless url
    
    if url.start_with?('/')
      "#{BASE_URL}#{url}"
    elsif url.start_with?('http')
      url
    else
      "#{BASE_URL}/#{url}"
    end
  end
  
  def save_article_to_file(article)
    filename = article[:title].gsub(/[^\w\s-]/, '').gsub(/\s+/, '_')[0..50]
    filepath = "#{OUTPUT_DIR}/articles/#{filename}.json"
    
    File.write(filepath, JSON.pretty_generate(article))
  end
  
  def save_summary
    summary = {
      total_articles: @scraped_articles.length,
      scraping_completed_at: Time.now.iso8601,
      total_words: @scraped_articles.sum { |a| a[:word_count] },
      articles_by_year: @scraped_articles.group_by { |a| a[:date] ? Date.parse(a[:date]).year : 'unknown' }.transform_values(&:count)
    }
    
    File.write("#{OUTPUT_DIR}/scraping_summary.json", JSON.pretty_generate(summary))
    puts "\nSummary saved to #{OUTPUT_DIR}/scraping_summary.json"
  end
  
  def save_failed_urls
    File.write("#{OUTPUT_DIR}/failed_urls.txt", @failed_urls.join("\n"))
    puts "Failed URLs saved to #{OUTPUT_DIR}/failed_urls.txt"
  end
end

# CLI interface
if __FILE__ == $0
  puts "Slovak Blog Scraper"
  puts "==================="
  
  # Check for credentials as command line arguments
  username = ARGV[0]
  password = ARGV[1]
  
  if username && password
    puts "Using provided credentials for authentication"
    scraper = BlogScraper.new(username, password)
  else
    puts "No credentials provided. Set them as arguments: ruby blog_scraper.rb username password"
    puts "Continuing without authentication (public content only)"
    scraper = BlogScraper.new
  end
  
  scraper.scrape_all
end

File: scripts/data_processing/blog_scraper.py
#!/usr/bin/env python3
"""
Slovak Blog Scraper
Scrapes articles from jaroslavlachky.sk blog.
"""

import requests
from bs4 import BeautifulSoup
import json
import os
import sys
import time
from datetime import datetime
from pathlib import Path
from urllib.parse import urljoin, urlparse
import argparse
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class BlogScraper:
    BASE_URL = 'https://jaroslavlachky.sk'
    BLOG_URL = f'{BASE_URL}/blog/'
    OUTPUT_DIR = './data/raw/scraped_data'
    
    def __init__(self):
        self.scraped_articles = []
        self.failed_urls = []
        
        # Setup session with proper headers
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        })
        
        self.setup_output_directory()
    
    def setup_output_directory(self):
        """Create output directories"""
        Path(self.OUTPUT_DIR).mkdir(parents=True, exist_ok=True)
        Path(f"{self.OUTPUT_DIR}/articles").mkdir(parents=True, exist_ok=True)
    
    def fetch_url(self, url):
        """Fetch URL content with error handling"""
        try:
            response = self.session.get(url, timeout=10)
            if response.status_code == 200:
                return response.text
            else:
                logger.warning(f"HTTP {response.status_code} for {url}")
                return None
        except Exception as e:
            logger.error(f"Network error for {url}: {e}")
            return None
    
    def normalize_url(self, url):
        """Normalize relative URLs to absolute URLs"""
        if not url:
            return None
        
        if url.startswith('/'):
            return f"{self.BASE_URL}{url}"
        elif url.startswith('http'):
            return url
        else:
            return f"{self.BASE_URL}/{url}"
    
    def scrape_article_urls_from_page(self, page_num):
        """Scrape article URLs from a blog page"""
        url = self.BLOG_URL if page_num == 1 else f"{self.BLOG_URL}page/{page_num}/"
        
        try:
            response = self.fetch_url(url)
            if not response:
                return []
            
            soup = BeautifulSoup(response, 'html.parser')
            
            # Try multiple selectors to find article links
            selectors = [
                'article h2 a',
                '.entry-title a',
                '.post-title a',
                'h2.entry-title a',
                '.article h2 a'
            ]
            
            article_links = []
            for selector in selectors:
                links = soup.select(selector)
                if links:
                    article_links = [self.normalize_url(link.get('href')) for link in links]
                    break
            
            # Fallback: look for any links that seem like blog posts
            if not article_links:
                all_links = soup.find_all('a', href=True)
                article_links = [
                    self.normalize_url(link['href'])
                    for link in all_links
                    if link['href'] and 
                       self.BASE_URL in link['href'] and 
                       '/blog/' not in link['href'] and 
                       link['href'].count('/') >= 4
                ]
                article_links = list(set(article_links))  # Remove duplicates
            
            return list(set(article_links))  # Remove duplicates
            
        except Exception as e:
            logger.error(f"Error fetching page {page_num}: {e}")
            return []
    
    def extract_title(self, soup):
        """Extract article title"""
        selectors = [
            'h1.entry-title',
            '.entry-title',
            'h1',
            'title'
        ]
        
        for selector in selectors:
            element = soup.select_one(selector)
            if element:
                title = element.get_text().strip()
                if title:
                    return title
        
        return None
    
    def extract_content(self, soup):
        """Extract article content"""
        selectors = [
            '.pf-content .entry_content',
            '.blog_entry_content',
            '.entry_content',
            '.post-content',
            '.article-content',
            '.content',
            'article .entry-content',
            'main article'
        ]
        
        for selector in selectors:
            element = soup.select_one(selector)
            if element:
                # Remove unwanted elements
                unwanted_selectors = [
                    'script', 'style', 'nav', 'header', 'footer', 'aside',
                    '.comments', '.social-share', '.in_share_element',
                    '.fb-like', '.twitter-like', '.printfriendly',
                    '.ve_form_element', 'form', '.mw_social_icons_container',
                    '.related_posts'
                ]
                
                for unwanted in unwanted_selectors:
                    for tag in element.select(unwanted):
                        tag.decompose()
                
                content = element.get_text().strip()
                if len(content) > 100:  # Minimum content length
                    return content
        
        # Fallback: try to get main content area
        main_selectors = ['main', '.main', '#main', '#content', '.content-area', '.blog-content']
        for selector in main_selectors:
            main_content = soup.select_one(selector)
            if main_content:
                # Remove unwanted elements
                unwanted_selectors = [
                    'script', 'style', 'nav', 'header', 'footer', 'aside',
                    '.comments', '.social-share', '.in_share_element',
                    '.fb-like', '.twitter-like', '.printfriendly', 'form'
                ]
                
                for unwanted in unwanted_selectors:
                    for tag in main_content.select(unwanted):
                        tag.decompose()
                
                content = main_content.get_text().strip()
                if len(content) > 100:
                    return content
        
        return ''
    
    def extract_date(self, soup):
        """Extract article date"""
        selectors = [
            'time[datetime]',
            '.published',
            '.entry-date',
            '.post-date',
            '[class*="date"]'
        ]
        
        for selector in selectors:
            element = soup.select_one(selector)
            if element:
                date_str = element.get('datetime') or element.get_text().strip()
                try:
                    # Try to parse the date
                    parsed_date = datetime.fromisoformat(date_str.replace('Z', '+00:00'))
                    return parsed_date.isoformat()
                except:
                    try:
                        # Fallback date parsing
                        from dateutil import parser
                        parsed_date = parser.parse(date_str)
                        return parsed_date.isoformat()
                    except:
                        continue
        
        return None
    
    def extract_excerpt(self, soup):
        """Extract article excerpt"""
        selectors = [
            '.excerpt',
            '.entry-summary',
            '.post-excerpt'
        ]
        
        for selector in selectors:
            element = soup.select_one(selector)
            if element:
                return element.get_text().strip()
        
        # Fallback: first paragraph of content
        first_p = soup.select_one('.entry-content p, .post-content p, .content p')
        if first_p:
            return first_p.get_text().strip()[:200]
        
        return ''
    
    def scrape_article(self, url):
        """Scrape a single article"""
        response = self.fetch_url(url)
        if not response:
            return None
        
        soup = BeautifulSoup(response, 'html.parser')
        
        # Extract title
        title = self.extract_title(soup)
        if not title:
            return None
        
        # Extract content
        content = self.extract_content(soup)
        if not content:
            return None
        
        # Extract metadata
        date = self.extract_date(soup)
        excerpt = self.extract_excerpt(soup)
        
        return {
            'url': url,
            'title': title,
            'content': content,
            'date': date,
            'excerpt': excerpt,
            'scraped_at': datetime.now().isoformat(),
            'word_count': len(content.split())
        }
    
    def save_article_to_file(self, article):
        """Save article to individual JSON file"""
        # Create safe filename
        safe_title = ''.join(c for c in article['title'] if c.isalnum() or c in (' ', '-', '_')).rstrip()
        safe_title = '_'.join(safe_title.split())[:50]
        filename = f"{safe_title}.json"
        filepath = Path(self.OUTPUT_DIR) / 'articles' / filename
        
        # Ensure unique filename
        counter = 1
        while filepath.exists():
            name_parts = safe_title.split('_')
            if name_parts[-1].isdigit():
                name_parts[-1] = str(counter)
            else:
                name_parts.append(str(counter))
            filename = f"{'_'.join(name_parts)[:50]}.json"
            filepath = Path(self.OUTPUT_DIR) / 'articles' / filename
            counter += 1
        
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(article, f, indent=2, ensure_ascii=False)
    
    def save_summary(self):
        """Save scraping summary"""
        # Group articles by year
        articles_by_year = {}
        for article in self.scraped_articles:
            if article.get('date'):
                try:
                    year = datetime.fromisoformat(article['date'].replace('Z', '+00:00')).year
                    articles_by_year[year] = articles_by_year.get(year, 0) + 1
                except:
                    articles_by_year['unknown'] = articles_by_year.get('unknown', 0) + 1
            else:
                articles_by_year['unknown'] = articles_by_year.get('unknown', 0) + 1
        
        summary = {
            'total_articles': len(self.scraped_articles),
            'scraping_completed_at': datetime.now().isoformat(),
            'total_words': sum(article['word_count'] for article in self.scraped_articles),
            'articles_by_year': articles_by_year
        }
        
        summary_path = Path(self.OUTPUT_DIR) / 'scraping_summary.json'
        with open(summary_path, 'w', encoding='utf-8') as f:
            json.dump(summary, f, indent=2, ensure_ascii=False)
        
        logger.info(f"Summary saved to {summary_path}")
    
    def save_failed_urls(self):
        """Save failed URLs to file"""
        if self.failed_urls:
            failed_path = Path(self.OUTPUT_DIR) / 'failed_urls.txt'
            with open(failed_path, 'w', encoding='utf-8') as f:
                f.write('\n'.join(self.failed_urls))
            logger.info(f"Failed URLs saved to {failed_path}")
    
    def scrape_all(self):
        """Main scraping function"""
        logger.info(f"Starting to scrape {self.BLOG_URL}")
        
        page = 1
        total_scraped = 0
        
        while True:
            logger.info(f"\n--- Scraping page {page} ---")
            article_urls = self.scrape_article_urls_from_page(page)
            
            if not article_urls:
                logger.info("No more articles found. Stopping.")
                break
            
            logger.info(f"Found {len(article_urls)} articles on page {page}")
            
            for url in article_urls:
                try:
                    article_data = self.scrape_article(url)
                    if article_data:
                        self.scraped_articles.append(article_data)
                        total_scraped += 1
                        logger.info(f"✓ Scraped: {article_data['title'][:50]}...")
                        
                        # Save individual article
                        self.save_article_to_file(article_data)
                        
                        # Small delay to be respectful
                        time.sleep(0.5)
                    else:
                        logger.warning(f"✗ Failed to scrape: {url}")
                        self.failed_urls.append(url)
                
                except Exception as e:
                    logger.error(f"✗ Error scraping {url}: {e}")
                    self.failed_urls.append(url)
            
            page += 1
            
            # Safety break - adjust if needed
            if page > 50:
                break
        
        logger.info("\n=== Scraping Complete ===")
        logger.info(f"Total articles scraped: {total_scraped}")
        logger.info(f"Failed URLs: {len(self.failed_urls)}")
        
        self.save_summary()
        if self.failed_urls:
            self.save_failed_urls()

def main():
    """CLI interface"""
    parser = argparse.ArgumentParser(description='Slovak Blog Scraper for jaroslavlachky.sk')
    parser.add_argument('--output', '-o', default='./data/raw/scraped_data', 
                       help='Output directory (default: ./data/raw/scraped_data)')
    
    args = parser.parse_args()
    
    print("Slovak Blog Scraper")
    print("===================")
    
    # Update output directory if specified
    if args.output:
        BlogScraper.OUTPUT_DIR = args.output
    
    scraper = BlogScraper()
    
    scraper.scrape_all()

if __name__ == "__main__":
    main()